dependency_cache.go

```diff
@@ -153,22 +150,46 @@ func customizeHttpClientTimeouts() (*HttpClientTimeouts, error) {
 	}, nil
 }
 
-// Returns the URI of a dependency mirror (optional).
-// Such mirror location can be defined in a binding of type 'dependency-mirror' with filename 'uri'
-// or using the environment variable 'BP_DEPENDENCY_MIRROR'. The latter takes precedence in case both are found.
-func getDependencyMirror(bindings libcnb.Bindings) (string, error) {
-	dependencyMirror := sherpa.GetEnvWithDefault("BP_DEPENDENCY_MIRROR", "")
-	// If no mirror was found in environment variables, try to find one in bindings.
-	if dependencyMirror == "" {
-		dependencyMirrorBindings, err := filterBindingsByType(bindings, "dependency-mirror")
-		if err == nil {
-			// Use the content of the file named "uri" as the mirror's URI.
-			dependencyMirror = dependencyMirrorBindings["uri"]
-		} else {
-			return "", err
+// Returns a key/value map with the URIs of all dependency mirrors. An empty map is returned if no mirrors are set.
+// Mirror locations can be defined in bindings of type 'dependency-mirror' or using env variables prefixed with 'BP_DEPENDENCY_MIRROR'.
+// Settings provided by env variables override those defined in bindings.
+func getDependencyMirrors(bindings libcnb.Bindings) (map[string]string, error) {
+	dependencyMirrors, err := filterBindingsByType(bindings, "dependency-mirror")
+	if err != nil {
+		return nil, err
+	}
+	dependencyMirrorsFromEnv := getDependencyMirrorsFromEnv()
+	for host, uri := range dependencyMirrorsFromEnv {
+		dependencyMirrors[host] = uri
+	}
+	return dependencyMirrors, nil
+}
+
+// Returns a key/value map of all dependency mirrors set in environment variables.
+func getDependencyMirrorsFromEnv() map[string]string {
+	mirrors := map[string]string{}
+	envs := os.Environ()
+	for _, env := range envs {
+		envPair := strings.SplitN(env, "=", 2)
+		hostnameSuffix, isMirror := strings.CutPrefix(envPair[0], "BP_DEPENDENCY_MIRROR")
+		hostnameEncoded, _ := strings.CutPrefix(hostnameSuffix, "_")
+		if isMirror {
+			mirrors[decodeHostnameEnv(hostnameEncoded)] = envPair[1]
 		}
 	}
-	return dependencyMirror, nil
+	return mirrors
+}
+
+// Takes an encoded hostname (from env key) and returns the decoded version in lower case.
+// Replaces double underscores (__) with one dash (-) and single underscores (_) with one period (.).
+func decodeHostnameEnv(encodedHostname string) string {
+	var decodedHostname string
+	if encodedHostname == "" {
+		decodedHostname = "default"
+	} else {
+		decodedHostname = strings.ReplaceAll(strings.ReplaceAll(encodedHostname, "__", "-"), "_", ".")
```

Comment:
Should check for characters outside of the ones that are allowed? Like `-`'s or `.`'s or even lowercase letters before we convert. I don't think we need to fail if there are invalid characters (although maybe if it's characters outside of what's allowed in a hostname?), because somehow those made it through the shell to us. If they make it through that's OK, but shells are weird and maybe this works on a dev's laptop but then not in CI. That's just a weird problem to troubleshoot. If we output some sort of warning in this situation, it would be visible and would train users to do it the way that should consistently work.

We could log a warning like "You have invalid characters in your mirror host environment variable. These will be allowed but for best results across different shells, you should replace `.` characters with `_` characters and `-` characters with `__`, and use all upper case letters. The buildpack will convert these back before using the mirror."

Reply:
Has been added.